Adds position and velocity motion generators to FrankaHWSim

rickstaa opened the pull request:
This pull request adds the `hardware_interface/PositionJointInterface'`and `hardware_interface/VelocityJointInterface` interfaces to the `FrankaHWSim` module. When a user chooses these interfaces, motion generators (i.e. PID controllers) are used for controlling the robot.

- This pull request requires #157 to be merged first.
- I did not extensively tune the PID gains of the motion generators, but they seem to work. Feel free to adjust them further.
- There is one minor bug in the current implementation that you might want to solve in future versions. The arm moves slightly at the beginning of the simulation when the position and velocity motion generators are used. This is probably because the joint commands are not initialized to the starting joint angles but to `0`. As I'm a bit short on time, I did not investigate this issue further. #181 is, in my opinion, good enough for releasing the Noetic version of the panda_moveit_config.

### TODOS

- [ ] Tune PID gains of the motion generators if you want smoother behaviour.
- [ ] Adjust the  trajectory controller goal tolerance  (see https://github.com/frankaemika/franka_ros/issues/201).

### Other TODOS
- [ ] Update documentation.

rhaschke commented on franka_description/robots/panda_arm.urdf.xacro:
```diff
@@ -44,13 +46,13 @@
-    <xacro:gazebo-joint joint="${arm_id}_joint5" transmission="hardware_interface/EffortJointInterface" />
-    <xacro:gazebo-joint joint="${arm_id}_joint6" transmission="hardware_interface/EffortJointInterface" />
-    <xacro:gazebo-joint joint="${arm_id}_joint7" transmission="hardware_interface/EffortJointInterface" />
+    <xacro:gazebo-joint joint="${arm_id}_joint1" transmission="${transmission}" />
+    <xacro:gazebo-joint joint="${arm_id}_joint2" transmission="${transmission}" />
+    <xacro:gazebo-joint joint="${arm_id}_joint3" transmission="${transmission}" />
+    <xacro:gazebo-joint joint="${arm_id}_joint4" transmission="${transmission}" />
+    <xacro:gazebo-joint joint="${arm_id}_joint5" transmission="${transmission}" />
+    <xacro:gazebo-joint joint="${arm_id}_joint6" transmission="${transmission}" />
+    <xacro:gazebo-joint joint="${arm_id}_joint7" transmission="${transmission}" />
```
You could use `$(arg transmission)` directly here, thus avoiding the intermediate property.

rickstaa replied:
@rhaschke Good catch, i will change that.

rhaschke commented on franka_gazebo/config/franka_hw_sim.yaml:
```diff
@@ -10,3 +10,26 @@ franka_gripper:
+# Motion generators PID gains
+# NOTE: Needed since the gazebo_ros_control currently only directly support effort hardware
+# interfaces (see #161).
+motion_generators:
+  position:
+    gains:
+      panda_joint1: { p: 600, d: 30, i: 0 }
+      panda_joint2: { p: 600, d: 30, i: 0 }
+      panda_joint3: { p: 600, d: 30, i: 0 }
+      panda_joint4: { p: 600, d: 30, i: 0 }
+      panda_joint5: { p: 250, d: 10, i: 0 }
+      panda_joint6: { p: 150, d: 10, i: 0 }
+      panda_joint7: { p:  50, d:  5, i: 0 }
+  velocity:
+    gains:
+      panda_joint1: { p: 600, d: 30, i: 0 }
+      panda_joint2: { p: 600, d: 30, i: 0 }
+      panda_joint3: { p: 600, d: 30, i: 0 }
+      panda_joint4: { p: 600, d: 30, i: 0 }
+      panda_joint5: { p: 250, d: 10, i: 0 }
+      panda_joint6: { p: 150, d: 10, i: 0 }
+      panda_joint7: { p:  50, d:  5, i: 0 }
```
These seem to be redundant to `effort_joint_trajectory_controller/gains` in [`sim_controllers.yaml`](https://github.com/frankaemika/franka_ros/pull/181/files#diff-62c417255cff239df1496453c784dd8326aa6654999d045613ce6986cb24bf37).

rickstaa replied:
@rhaschke Thanks for your comment. The PID gains that are found above are similar to the `gazebo_ros_control.pid_gains` that are used in the [DefaultRobotHWSim](https://github.com/ros-controls/gazebo_ros_control/blob/46f41d0b8b90e4b29c2500123d879e768bba830c/src/default_robot_hw_sim.cpp#L203) of the [gazebo_ros_control](https://github.com/ros-controls/gazebo_ros_control) package.

These gains are only used when the user requests a `position` or `velocity` hardware_interface:

https://github.com/frankaemika/franka_ros/blob/f9f3c9147f73a18c73c5d2058678b1413b309dc9/franka_gazebo/src/franka_hw_sim.cpp#L115-L139

When a `effort` hardware_interface is requested these effort commands are directly forwarded to the [joint->setForce()](https://osrf-distributions.s3.amazonaws.com/gazebo/api/dev/classgazebo_1_1physics_1_1Joint.html#ab2491053d1c5ebb97c377064797af494) method:

https://github.com/frankaemika/franka_ros/blob/f9f3c9147f73a18c73c5d2058678b1413b309dc9/franka_gazebo/src/franka_hw_sim.cpp#L402

Although, in the current version I used the exact same gains for the simulated [motion generators](https://github.com/rickstaa/franka_ros/blob/f9f3c9147f73a18c73c5d2058678b1413b309dc9/franka_gazebo/config/sim_controllers.yaml#L69-L86) and the `effort_joint_trajectory_controller` they are to my knowledge distinct. 

It however is true that for the release of the [panda_moveit_config](https://github.com/ros-planning/panda_moveit_config) the `effort_joint_trajectory_controller` is not needed. I added it to this pull request to give users that want to use an effort based trajectory controller instead of the position based on a starting point (see https://github.com/frankaemika/franka_ros/issues/161#issuecomment-930173703).

rickstaa replied:
@rhaschke Please let me know if I misunderstood your comment. I think for now we can leave the decision to include the `effort_joint_trajectory_controller` in the `franka_gazebo/config/sim_controllers.yaml` file to the @frankaemika team. I think, if it is included in the `sim_controllers.yaml` file, it should also be included in the [franka_control/config/default_controllers.yaml](https://github.com/frankaemika/franka_ros/blob/develop/franka_control/config/default_controllers.yaml) file.

rickstaa replied:
@rhaschke I removed the `effort_joint_trajectory_controller` as the PID gains do not seem to work on the real robot. MoveIt more than often times out. I think this is because no integral gain is used.

rhaschke reviewed (commented):
Just a few remarks. Generally looks good to me.

rickstaa commented:
@gollth I think this pull request provides the best solution since users can now work with the simulation in the same way as they work with the real robot. I think it is ready to be merged after we improved the PID gains of the motion planners (see GIF below). 

![pid_gains_improve](https://user-images.githubusercontent.com/17570430/138105144-62aedb9e-1cdd-4e85-b3f2-2abd846ccad9.gif)

rickstaa commented:
As explained in #135, I based the code in this pull request on the implementation found in the [gazebo_ros_control](https://github.com/ros-simulation/gazebo_ros_pkgs/pull/135). There they also use the `control_toolbox` PID's instead of the PID controllers that are implemented by Gazebo (i.e. [SetPositionTarget ](https://osrf-distributions.s3.amazonaws.com/gazebo/api/dev/classgazebo_1_1physics_1_1JointController.html#a5b89ea86a2df7638009f6704f0f14f90) and [SetVelocityTarget](https://osrf-distributions.s3.amazonaws.com/gazebo/api/dev/classgazebo_1_1physics_1_1JointController.html#a3113f99b3ae4675e133a1be217521d6d)). The reason is given [here](https://github.com/ros-simulation/gazebo_ros_pkgs/pull/135#issuecomment-32350640):

> > Gazebo already has built-in PID controllers for these two control modes
> 
> I believe that the reason why control_toolbox's PIDs are used instead of Gazebo's implementation is that loading the controller configuration (gains and clampings) is straightforward and already implemented for the former (via ROS params).

For the `FrankaHWSim` we could therefore also directly use the Gazebo PID controllers. This will, however, make tuning a bit harder.

rickstaa commented:
Closed in favor of #183

rhaschke commented:
I'm confused. #183 doesn't deal with motion generators in FrankaHWSim. However, the contributions here are required for `[position|velocity]_controllers/JointTrajectoryController` to be used in simulation, aren't they?

rickstaa commented:
@rhaschke Ah your right! Based on our meeting I assumed your pull request was fully based on #181. I now see that it only includes the commits from #157.

rickstaa commented:
@gollth I found the comment from when I tested the effort controller PID gains (see https://github.com/frankaemika/franka_ros/pull/181#discussion_r734644377). I however did not provide much information other than that I received a `ros_control` timeout. As a result MoveIt also timed out. I do remember that the arm did not move. I can check the exact message on Monday.

rickstaa commented:
@rhaschke Do you want me to include the commits of #157 into this pull request?

rhaschke commented:
> Do you want me to include the commits of #157 into this pull request?

No, they are related but provide rather independent functionality.

rickstaa commented:
@gollth I removed https://github.com/frankaemika/franka_ros/pull/181#issuecomment-953022098 since the simulation gains you provided also seem to work on the real robot (see https://github.com/frankaemika/franka_ros/pull/186#issuecomment-956144658).

rickstaa committed:
feat: add effort and position traj controllers to simulation

rickstaa commented on franka_gazebo/config/sim_controllers.yaml:
```diff
@@ -1,3 +1,51 @@
+    - panda_joint7
+  constraints:
+    goal_time: 0.5
+    panda_joint1: { goal: 0.05}
```
Here I used the trajectory controller goal tolerance that is found in the `franka_control` package. I however think this is too big `https://github.com/frankaemika/franka_ros/issues/201`. Please update the `franka_control` package so that I can update it here.

rickstaa commented on franka_gazebo/config/sim_controllers.yaml:
```diff
@@ -1,3 +1,51 @@
+    panda_joint7: { p: 50, d: 5, i: 0 }
+  constraints:
+    goal_time: 0.5
+    panda_joint1: { goal: 0.05}
```
Same story.

xianyitiantim commented:
Hi @rickstaa , could you please provide any launch file on how to use this merged package so that the joint position could be controlled? Thanks!

rickstaa commented:
@Yitian-Xian Sorry for the late reply. You should be able to control the simulated robot using position and velocity control by merging both #181 and #157 into the [develop branch](https://github.com/frankaemika/franka_ros/tree/develop) and build [franka_ros from the source](https://frankaemika.github.io/docs/installation_linux.html). After that, you can launch the joint position-controlled Gazebo robot using the `roslaunch franka_gazebo panda.launch transmission:=position` command. The gazebo panda robot can then be controlled with the [ros_control position controllers](http://wiki.ros.org/ros_control). I don't have a small example repository right now but you can check my code in which I use the position control (see https://github.com/rickstaa/panda-gazebo/blob/fdf45991f5e04569abbb590bd4fe01c438610fde/panda_gazebo/launch/put_robot_in_world.launch#L26). Let me know if you run into problems.

rickstaa commented:
@Yitian-Xian To save you some time, I quickly created an example branch to test out the new position motion generators. To test them out, please follow the following steps:

1. Build libfranka from source according to [the documentation](https://frankaemika.github.io/docs/installation_linux.html).
2. Create a catkin workspace.
3. Clone the example [rickstaa/franka_ros into/position_control_example](https://github.com/rickstaa/franka_ros/tree/position_control_example) this workspace.
4. Install the ros dependencies using `rosdep install --from-path src --ignore-src -y -r --skip-keys libfranka
`.
5. Build the catkin workspace `catkin build -DCMAKE_BUILD_TYPE=Debug -DFranka_DIR:PATH=<PATH_TO_INSTALL>/libfranka/build`.
6. Source the workspace.
7. Start the panda simulation while setting the `transmission` to position:

```bash
roslaunch franka_gazebo panda.launch transmission:=position
```

8. Load the position controllers:

```bash
roslaunch franka_gazebo load_joint_position_controllers.launch
```

9. **OPTIONAL:** Load the [rqt_controller_manager](http://wiki.ros.org/rqt_controller_manager) to check if the controllers are running:

```bash
rosrun rqt_controller_manager rqt_controller_manager
```

10. Start the position control dynamic reconfigure server:

```bash
 rosrun franka_gazebo joint_position_dynamic_reconfigure_server.py
```

11. Start the dynamic reconfigure GUI:

```bash
rosrun rqt_gui rqt_gui -s reconfigure
```

12. Control the robot by changing the joint positions in the `joint_position_reconfig_server` tab.
13. See that it works.

## PID gains

Please note that I did not extensively tune the PID gains as stated above. If you want to use the motion generators for your experiments, you have to improve the PID gains. This can be done by modifying the PID gains under the `motion_generators/position/gains/` tab in the dynamic reconfigure GUI. 

![image](https://user-images.githubusercontent.com/17570430/150646640-be7e675b-bf7a-4058-9b41-ac3e8ee236ec.png)

Might you find better PID gains? Can you maybe share them here or create a pull request on https://github.com/rickstaa/franka_ros/tree/add_motion_generators to include them for everybody to use. 🚀 

If you have any questions or run into problems, please let me know.

rickstaa commented:
@fwalch, @marcbone Did you already conclude whether you want to merge this into the develop branch or that we first need to make some changes?

rtkg commented:
I would add my voice for merging this PR. We're currently running a custom Gazebo hardware interface to allow position/velocity control which is essential to have a reasonable automated testing pipeline.

rpapallas commented:
We are also having the same issue, and we would like to have the position/velocity controllers working, please.

rickstaa committed:
feat: add position and velocity motion generators to FrankaHWSim

This commit adds the 'hardware_interface/PositionJointInterface' and
'hardware_interface/VelocityJointInterface' interfaces to the
FrankaHWSim module. When a user chooses these interfaces, motion
generators (i.e. PID controllers) are used for controlling the robot.

rickstaa commented on franka_gazebo/src/franka_hw_sim.cpp:
```diff
@@ -480,11 +558,9 @@ void FrankaHWSim::updateRobotState(ros::Time time) {
-    this->robot_state_.q_d[i] = joint->position;  // special case for resetting motion generators
-    this->robot_state_.dq_d[i] = 0;
-    this->robot_state_.ddq_d[i] = 0;
+    this->robot_state_.q_d[i] = joint->position;
```
@gollth Since there is a lot of interest in this pull request, I again rebased it on the `develop` branch. I also fixed the conflicts that were introduced due to https://github.com/frankaemika/franka_ros/pull/198 being merged. Here I undid some changes made in  https://github.com/frankaemika/franka_ros/pull/198 since these modes are also supported after #181 is merged. You can of course, also add some conditional logic using the `joint_control_methods_[joint->name]` variable. I will leave this to you.

rpapallas commented:
Thank you, @rickstaa, for your changes, I have merged them into our fork.

rickstaa commented:
@rpapallas No problem. Glad I could help. I also created a pr that combines all the changes needed for the position and motion generators to work see https://github.com/frankaemika/franka_ros/pull/224. Please be aware that the used PID gains are not extensively tuned. I currently am not using position and velocity control myself. Might you and your team find better gains feel free to create a pull request onto #181.

rpapallas commented:
Thank you, I have merged this as well. Thank you for the heads-up; we will just for now use this in simulation and wait for Franka to officially merge these, hopefully soon.

rpapallas commented:
Hi @rickstaa although I merged both changes, I am getting the following error when I start franka gazebo with the joint position controller:

```
INFO] [1644924385.485943, 0.150000]: Loading controller: joint_position_example_controller
[ERROR] [1644924385.490905630, 0.153000000]: Could not load controller 'joint_position_example_controller' because the type was not specified. Did you load the controller configuration on the parameter server (namespace: '/joint_position_example_controller')?
[panda_model_spawner-4] process has finished cleanly
```

Did you have similar issues in the past?

rickstaa commented:
@rpapallas The pull request found in #224 do currently not configure any position/effort controllers. They only add the required code changes to use position controllers. You are still expected to load your own position/velocity controllers. For an example, see https://github.com/frankaemika/franka_ros/pull/181/commits/e60cf080020b96e94d532f3476ccb6ad0f99eaff. Also see https://github.com/frankaemika/franka_ros/pull/181#issuecomment-1019301350 for all the steps required for using the new (simulated) position and velocity motion generators. 

I guess @frankaemika could include a position and velocity example controller in the https://github.com/frankaemika/franka_ros/blob/develop/franka_gazebo/config/sim_controllers.yaml. I, however, did not do this since the position and effort motion generators are not 100% equal to the motion generators on the real robot. I, therefore, left this decision to @frankaemika. 

I did include a `position_joint_trajectory_controller` because this controller is required by [MoveIt tutorials](https://ros-planning.github.io/moveit_tutorials/) and these tutorials are often used in combination with the Panda robot.

https://github.com/frankaemika/franka_ros/blob/765cacda92584d9898da966c737fb6d310e0e3fc/franka_gazebo/config/sim_controllers.yaml#L70

rpapallas commented:
Thank you, @rickstaa. I was able to get it working.

marcbone commented:
Ok guys. We heavily modified this PR and merged it into develop.

The feature works now a bit different. You do not have to specify at launch time which type of interface you want to use. All three are available. So you can start a position controller, stop it and then run an effort controller. Just like on the real robot. However, you cannot have a controller to control only one joint as @rickstaa  did for his demo, where he had 7 controllers (one for each joint). You need to use one controller to control all joints, as you would on the real robot.
Also when you launch the robot in gazebo without any controller, the robot will now keep its position.

So any controller that does not use the cartesian pose or cartesian velocity interface and runs on the real robot should now also work in gazebo.

I am going to close this PR , #157 and #224

rpapallas commented:
Thank you for this; much appreciated.

rpapallas commented:
Thank you for this; much appreciated.

@marcbone, To get this simulated position controller working in gazebo, I use the following:
```
roslaunch franka_gazebo panda.launch controller:=joint_position_example_controller
```

But I can't seem to find where the controller is expecting the new joint positions, is there a ros topic / actionlib that I need to use to send joint positions to the robot?

Thank you.

gollth commented:
Hi @rpapallas,

you can for example start the [Position Joint Trajectory Controller](http://wiki.ros.org/joint_trajectory_controller) and use its [action](http://wiki.ros.org/joint_trajectory_controller#Action_interface) or [topic interface](http://wiki.ros.org/joint_trajectory_controller#Subscribed_Topics)

```console
$ roslaunch franka_gazebo panda.launch controller:=position_joint_trajectory_controller
``` 

```console
$ rostopic list /position_joint_trajectory_controller
/position_joint_trajectory_controller/command                                # <-  Topic interface
/position_joint_trajectory_controller/follow_joint_trajectory/cancel         # -.
/position_joint_trajectory_controller/follow_joint_trajectory/feedback       #  |
/position_joint_trajectory_controller/follow_joint_trajectory/goal           #  }- Action interface
/position_joint_trajectory_controller/follow_joint_trajectory/result         #  |
/position_joint_trajectory_controller/follow_joint_trajectory/status         # -`
/position_joint_trajectory_controller/state

```

rpapallas commented:
Hi @gollth thank you for the details. I don't think I can run that successfully despite having `ros-noetic-ros-controllers`, `ros-noetic-trajectory-controller` and `ros-noetic-joint-trajectory-controller` installed.

Running the controller spawner alone also suggests that it can't find the controller:
```
rosrun controller_manager spawner position_joint_trajectory_controller
```
with the following output:
```
[INFO] [1647252909.642155]: Controller Spawner: Waiting for service controller_manager/load_controller
[WARN] [1647252939.845601]: Controller Spawner couldn't find the expected controller_manager ROS interface.
```

marcbone commented:
Try to install `ros-noetic-position-controllers`, but actually they should be part of ros-controllers.

Does the controller show up in the `rqt_controller_manager`?

rpapallas commented:
Hi @marcbone indeed the position-controllers is installed. The `rqt_controller_manager` opens and has an empty dropdown menu when I run it through `rosrun rqt_controller_manager rqt_controller_manager`

marcbone commented:
That is strange. Could you post the output of `roslaunch franka_gazebo panda.launch` please?

gollth commented:
Sometimes RQT does not reload [newly installed plugins automatically](https://answers.ros.org/question/91231/rqt-plugin-not-listedfound-in-list-returned-by-rqt-list-plugins/). Does

```bash
rqt --force-discover
```

help?

rpapallas commented:
Hi @gollth thank you for your support. I found the problem. I am using a Singularity container to have the system working (it's similar to a Docker container). I installed ROS Noetic and Ubuntu 20.04 but only Python3. Your `delayed_controller_spawner.py`'s shebang points to Python 2 and it couldn't find it. Since Python 2 is not supported, I changed the shebang to `#!/usr/bin/env python3` and it now works.

rpapallas commented:
If you think that shebang should be Python 3, I can create a PR, if you want.

gollth commented:
Changing the shebang to python3 would make it so that `delayed_controller_spanwer.py` wouldn't work on melodic anymore.. I think the script is simple enough to work on both python2 and python3. Can you just install `python-is-python3` in your docker container?

rpapallas commented:
Perfect; didn't know about this! Thank you very much for the help!

rhaschke commented:
@gollth, you can install python scripts via `catkin_install_python()` as explained [here](http://wiki.ros.org/UsingPython3/SourceCodeChanges#Changing_shebangs) and have the shebang rewritten correctly.

gollth commented:
> have the shebang rewritten correctly.

In the link you sent it seems the correct way is `#!/usr/bin/env python` what we already have in the script. Could you elaborate what you mean with "correct" @rhaschke ?

rhaschke commented:
The cmake function `catkin_install_python()` will _"correctly"_ rewrite the shebang during the installation process to use either python2 or python3. Thus, `python-is-python3` or `python-is-python2` is not needed when running from an install space.
The question is how you install the python script(s). However, I couldn't find `delayed_controller_spanwer.py`.

marcbone commented:
@rhaschke Thanks for the hint. I fixed it

rpapallas commented:
Many new things learnt today! Glad I am in this thread 😊 thank you, @marcbone and @rhaschke!

rickstaa commented:
> Ok guys. We heavily modified this PR and merged it into develop.
> 
> The feature works now a bit different. You do not have to specify at launch time which type of interface you want to use. All three are available. So you can start a position controller, stop it and then run an effort controller. Just like on the real robot. However, you cannot have a controller to control only one joint as @rickstaa did for his demo, where he had 7 controllers (one for each joint). You need to use one controller to control all joints, as you would on the real robot. Also when you launch the robot in gazebo without any controller, the robot will now keep its position.
> 
> So any controller that does not use the cartesian pose or cartesian velocity interface and runs on the real robot should now also work in gazebo.
> 
> I am going to close this PR , #157 and #224

@marcbone Thanks for getting this implemented. Your implementation makes a lot of sense! I will perform some tests today to see which of my issues can be closed.

rickstaa commented:
# Test results

@marcbone I reviewed your pull request, and all of the controllers below seem to work as expected (see table below). Thanks for improving on my PR. 

|                                                 | Gazebo Tests | Comment                        | franka_control | franka_example_controllers | franka_gazebo | Loaded by default |
|-------------------------------------------------|--------------|--------------------------------|----------------|----------------------------|---------------|-------------------|
| joint_velocity_example_controller               | ✔️            |                                | ❌              | ✔️                          | ✔️             | ✔️                 |
| joint_position_example_controller               | ✔️            |                                | ❌              | ✔️                          | ✔️             | ✔️                 |
| cartesian_pose_example_controller               | ❌            | Interface not yet implemented. | ❌              | ✔️                          | ❌             | ❌                 |
| elbow_example_controller                        | ❌            | Interface not yet implemented. | ❌              | ✔️                          | ❌             | ❌                 |
| cartesian_velocity_example_controller           | ❌            | Interface not yet implemented. | ❌              | ✔️                          | ❌             | ❌                 |
| model_example_controller                        | ✔️            |                                | ❌              | ✔️                          | ✔️             | ✔️                 |
| force_example_controller                        | ✅            | Works but drift due to #160.   | ❌              | ✔️                          | ✔️             | ✔️                 |
| joint_impedance_example_controller              | ❌            | Interface not yet implemented. | ❌              | ✔️                          | ❌             | ❌                 |
| cartesian_impedance_example_controller          | ✔️            |                                | ❌              | ✔️                          | ✔️             | ✔️                 |
| dual_arm_cartesian_impedance_example_controller | -            | Not tested.                    | ❌              | ✔️                          | ❌             | ❌                 |
| teleop_joint_pd_example_controller              | -            | Not tested.                    | ❌              | ✔️                          | ❌             | ❌                 |
| franka_state_controller                         | ✔️            |                                | ✔️              | ❌                          | ✔️             | ✔️                 |
| position_joint_trajectory_controller            | ✔️          |                                | ✔️              | ❌                          | ✔️             | ✔️                 |
| effort_joint_trajectory_controller              | ✔️          |                                | ✔️              | ❌                          | ✔️             | ✔️                 |
| joint_effort_controller                         | ✔️             |                                | ❌              | ❌                          | ❌             | ❌                 |
| joint_position_controller                       |  ✔️            |                                | ❌              | ❌                          | ❌             | ❌                 |
| joint_velocity_controller                       | ✔️             |                                | ❌              | ❌                          | ❌             | ❌                 |
| franka_gripper                                  | ✔️            |                                | ❌              | ❌                          | ❌             | ✔️                 |

## Steps to test these controllers


1. Build libfranka from source according to [the documentation](https://frankaemika.github.io/docs/installation_linux.html).
2. Create a catkin workspace.
3. Clone [rickstaa/test-effort-position-velocity-controllers](https://github.com/rickstaa/franka_ros/tree/test-effort-position-velocity-controllers) into this workspace.
4. Install the ros dependencies using `rosdep install --from-path src --ignore-src -y -r --skip-keys libfranka
`.
5. Build the catkin workspace `catkin build -DCMAKE_BUILD_TYPE=Debug -DFranka_DIR:PATH=<PATH_TO_INSTALL>/libfranka/build`.
6. Source the workspace.
7. Start the panda simulation.

```bash
roslaunch franka_gazebo panda.launch 
```

8. Load the [rqt_controller_manager](http://wiki.ros.org/rqt_controller_manager) and start a position/effort/velocity/trajectory controller.

```bash
rosrun rqt_controller_manager rqt_controller_manager
```

9. Start a position/effort/velocity or trajectory dynamic reconfigure server:

```bash
 rosrun franka_gazebo joint_position_dynamic_reconfigure_server.py
```

```bash
rosrun rqt_joint_trajectory_controller rqt_joint_trajectory_controller
```

11. For the position/effort/velocity control also start the dynamic reconfigure GUI:

```bash
rosrun rqt_gui rqt_gui -s reconfigure
```

12. Control the robot by changing the joint positions/efforts/velocities in the `joint_position_reconfig_server` tab.
13. See that it works.

rickstaa commented:
@marcbone, @gollth I just reviewed that all new controllers work as expected (see https://github.com/frankaemika/franka_ros/pull/181#issuecomment-1068939918). I, however, have several minor remarks:
    - If people want to use simulated position controls in research, PID controller gains might still need to be improved. People, however, can do this in the future.
    - Gravity compensation is still broken when working with effort or velocity control (see #160). The current solution only masks the problem because position control is enabled by default. I will continue my discussion on this in #160.
    - Maybe it is an idea to add the configurations for the [ros_control](http://wiki.ros.org/ros_control) position/effort/velocity group controllers to the controller configurations (i.e. https://github.com/frankaemika/franka_ros/blob/develop/franka_gazebo/config/sim_controllers.yaml and https://github.com/frankaemika/franka_ros/blob/develop/franka_control/config/default_controllers.yaml). I think most people are going to use these controllers when they start instead of writing C++ controllers themselves.
 
```yaml
joint_position_group_example_controller:
  type: effort_controllers/JointGroupEffortController
  joint_names:
      - $(arg arm_id)_joint1
      - $(arg arm_id)_joint2
      - $(arg arm_id)_joint3
      - $(arg arm_id)_joint4
      - $(arg arm_id)_joint5
      - $(arg arm_id)_joint6
      - $(arg arm_id)_joint7
 ```
